### PR TITLE
[feat] handle errors for all auth flows

### PR DESCRIPTION
**Pull request by ccatherinetan**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
Files Edited with Error-handling functionality 
* `src/app/(auth)/login/page.tsx`
* `src/app/(auth)/signup/page.tsx`
* `src/app/(auth)/reset-password/page.tsx`
* `src/app/(auth)/forgot-password`
     * Error handling for Forgot Password (checks email validity before sending the reset password email)

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @davidqing6432 at the bottom."

<details>
<summary> Log In</summary>
Empty Entries 
Screen Shot 2023-11-27 at 6 11 33 PM

Invalid Credentials
Screen Shot 2023-11-27 at 6 11 45 PM
</details>

<details>
<summary> Sign Up </summary>
Empty Entries 
Screen Shot 2023-11-27 at 6 12 10 PM

Invalid Email
Screen Shot 2023-11-21 at 3 52 42 PM
</details>

<details>
<summary> Reset Password </summary>

Empty New Password
Screen Shot 2023-11-27 at 1 50 25 PM

Password Not Matching 
Screen Shot 2023-11-27 at 1 51 42 PM

Supabase Password Error 
Screen Shot 2023-11-27 at 1 51 36 PM
</details>

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
* `src/app/(auth)/login/page.tsx`
   * try invalid credentials (e.g. wrong password) → (should error)
   * try invalid email → (should error)
   * try invalid password (less than 6 chars) → (should error)
* `src/app/(auth)/signup/page.tsx`
   * try invalid email → (should error)
   * try invalid password (less than 6 chars) → (should error)
   * try to create a new account with an email that's already associated with an account → (should error)
* `src/app/(auth)/reset-password/page.tsx`
   * try invalid password (less than 6 chars) → (only "new password" should error)
   * try same password → (should error at the top)
   * try non-matching passwords → (should error at the top)
* `src/app/(auth)/forgot-password`
   * the error does not disappear until a valid email is entered and the button is reclicked. 

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
* finalizing password complexity requirements, besides length. (currently, the only password complexity requirement for the length to be >= 6). → update signup and reset-password w/ these password requirements 
* i haven't been able to implement the global background color
* for MVP/stretch: throw error if email is not in the profiles table (i.e. the email is not associated with an actual user) 
* reset-password: 

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @pragyakallanagoudar
@davidqing6432

### CODE REVIEW

File: src/app/(auth)/login/page.tsx [hunk 1]
```diff
@@ -45,13 +49,15 @@ export default function Login() {
           <TextInput
             label="Password"
             placeholder="Password"
-            errorText="" // "Password Error"
+            errorText={
+              password === '' || password.length < 6 ? 'Invalid Password' : ''
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/login/page.tsx]: (cc @davidqing6432) can we get rid of the `password.length < 6` check here? think that once they're at the point of logging in, we shouldn't be checking any password constraints

File: src/app/(auth)/login/page.tsx [hunk 2]
```diff
@@ -23,20 +23,24 @@ export default function Login() {
...
+        <H1>Log In</H1>
+        {errorMessage !== '' && <P $color={COLORS.redMid}>{errorMessage}</P>}
+      </SpacerDiv>
+      <SpacerDiv $gap={13 / 16}>
+        <SpacerDiv>
           <TextInput
             label="Email"
             placeholder="email@example.com"
-            errorText="" // "Email Error"
+            errorText={email === '' ? 'Invalid Email' : ''}
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/login/page.tsx]: (cc @davidqing6432) instead of immediately checking for potential empty email/password fields, can we only check for this when the "Log In" button is pressed? i think having the page load with error fields already in place might be a bit confusing.

File: src/app/(auth)/login/page.tsx [hunk 3]
```diff
@@ -23,20 +23,24 @@ export default function Login() {
...
   };
 
   return (
     <>
-      <H1>Log In</H1>
-      <SpacerDiv>
-        <QuestionsDiv>
+      <SpacerDiv $gap={0.625}>
+        <H1>Log In</H1>
+        {errorMessage !== '' && <P $color={COLORS.redMid}>{errorMessage}</P>}
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/login/page.tsx:49]: Can we make it so that the introduction of the error message does not shift around other components? (i.e. have it still take up the same amount of space when an error is not present?)

**Comment by jinkang-0** [comment]
we might want to change the phrasing of the error message for an invalid email address (colon feels a bit... odd)
maybe "Invalid email address"?

image

File: src/app/(auth)/login/page.tsx (hunk 3 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx:49]: Ahh @davidqing6432 what do you think? I think the current implementation is consistent with figma.

File: src/app/(auth)/reset-password/page.tsx [hunk 4]
```diff
@@ -22,29 +24,39 @@ export default function ResetPassword() {
...
+      // throw new Error(
+      //   `An error occurred trying to reset password: ${error.message}`,
+      // );
+    } else {
+      const { error: signOutError } = await supabase.auth.signOut();
+      if (signOutError) {
+        throw new Error(
+          `An error occurred trying to sign out: ${signOutError.message}`,
+        );
+      }
```
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx:42]: we should consider some other ways to handle error besides just throwing them since it will crash the application

File: src/app/(auth)/login/page.tsx (hunk 1 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx]: i can get rid of them!

File: src/app/(auth)/login/page.tsx (hunk 2 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx]: sounds good! yeah i agree that having them appear initially is confusing. 
do we want this change for the sign up page too?

File: src/app/(auth)/reset-password/page.tsx [hunk 5]
```diff
@@ -22,29 +24,37 @@ export default function ResetPassword() {
...
-        <QuestionsDiv>
+        <SpacerDiv $gap={0.625}>
+          <H1>Set New Password</H1>
+          {errorMessage !== '' && <P $color={COLORS.redMid}>{errorMessage}</P>}
+        </SpacerDiv>
+        <SpacerDiv>
           <TextInput
             label="New Password"
             placeholder="Password"
+            errorText={newPassword.length < 6 ? 'Invalid Password' : ''}
```
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx:58]: i think we should only set this when the user click on the button, since it could be weird to see an error while the user is still typing out their password

tho also a design question

File: src/app/(auth)/signup/page.tsx [hunk 6]
```diff
@@ -14,8 +14,28 @@ export default function SignUp() {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
   const [emailSentCount, setEmailSentCount] = useState(0);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [emailError, setEmailError] = useState('');
+  const [passwordError, setPasswordError] = useState('');
+
   // const { push } = useRouter();
   const handleSignUp = async () => {
+    if (email === '' || password === '') {
```
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: i don't think we need this bigger if statement if we're checking them again..

File: src/app/(auth)/reset-password/page.tsx [hunk 7]
```diff
@@ -3,15 +3,17 @@
...
 import BigButton from '@/components/BigButton';
 import supabase from '@/api/supabase/createClient';
 
 export default function ResetPassword() {
   const [newPassword, setNewPassword] = useState('');
   const [newPassword2, setNewPassword2] = useState('');
   const [canReset, setCanReset] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
   const { push } = useRouter();
   useEffect(() => {
```
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx:19]: nit: can we add a line before `useEffect` lol

File: src/app/(auth)/login/page.tsx (hunk 3 above)
Comment (pragyakallanagoudar) [reply src/app/(auth)/login/page.tsx:49]: oop i didn't even cross-reference with the figma, that's my bad! in that case we can leave this be :)

File: src/app/(auth)/reset-password/page.tsx (hunk 4 above)
Comment (pragyakallanagoudar) [reply src/app/(auth)/reset-password/page.tsx:42]: oh VERY true, that should definitely be included in this sprint 

@davidqing6432 thoughts? all i can think of is a toast notification

File: src/app/(auth)/reset-password/page.tsx (hunk 5 above)
Comment (pragyakallanagoudar) [reply src/app/(auth)/reset-password/page.tsx:58]: personally, i think this is chill bc ultimately it will be replaced with password complexity stuff -- which will show up as the user types in the password. but yes def design question, cc @davidqing6432

File: src/app/(auth)/reset-password/page.tsx (hunk 4 above)
Comment (ccatherinetan) [reply src/app/(auth)/reset-password/page.tsx:42]: yes definitely! i wasn't really sure what to do here, since i don't think a sign out error should occur. @davidqing6432 @pragyakallanagoudar thoughts on error handling?

File: src/app/(auth)/login/page.tsx [hunk 8]
```diff
@@ -7,36 +7,60 @@ import TextInput from '@/components/TextInput/index';
...
+      </SpacerDiv>
+      <SpacerDiv $gap={13 / 16}>
```
Comment (jinkang-0) [inline src/app/(auth)/login/page.tsx]: nit: can we turn this into a single number?

File: src/app/(auth)/signup/page.tsx [hunk 9]
```diff
@@ -14,8 +14,28 @@ export default function SignUp() {
...
+      if (email === '') {
+        setEmailError('Invalid Email');
+      } else {
+        setEmailError('');
+      }
```
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: nit: we can probably turn this into a ternary statement
i.e. `setEmailError((email === '') ? 'Invalid Email' : '')`

File: src/app/(auth)/signup/page.tsx [hunk 10]
```diff
@@ -14,8 +14,28 @@ export default function SignUp() {
...
+      if (password === '') {
+        setPasswordError('Invalid Password');
+      } else {
+        setPasswordError('');
+      }
```
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: nit: we can probably turn this into a ternary statement also

File: src/app/(auth)/signup/page.tsx [hunk 11]
```diff
@@ -14,8 +14,28 @@ export default function SignUp() {
...
+      setErrorMessage('');
+      return;
+    }
```
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx:31]: we can probably unwrap the bigger if statement and add it here
i.e. `if (email === '' || password === '') return;`

File: src/app/(auth)/signup/page.tsx (hunk 6 above)
Comment (ccatherinetan) [reply src/app/(auth)/signup/page.tsx:23]: i have this statement so that it returns out of the func before calling the supabase signup function in case either email or password is invalid.

File: src/app/(auth)/login/page.tsx (hunk 8 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx]: done!

File: src/app/(auth)/signup/page.tsx (hunk 9 above)
Comment (ccatherinetan) [reply src/app/(auth)/signup/page.tsx]: yes! done!

File: src/app/(auth)/signup/page.tsx (hunk 10 above)
Comment (ccatherinetan) [reply src/app/(auth)/signup/page.tsx]: yup!

File: src/app/(auth)/signup/page.tsx (hunk 11 above)
Comment (ccatherinetan) [reply src/app/(auth)/signup/page.tsx:31]: done!

File: src/app/(auth)/reset-password/page.tsx (hunk 7 above)
Comment (ccatherinetan) [reply src/app/(auth)/reset-password/page.tsx:19]: done!

File: src/app/(auth)/signup/page.tsx (hunk 6 above)
Comment (ccatherinetan) [reply src/app/(auth)/signup/page.tsx]: gotcha!

File: src/app/(auth)/reset-password/page.tsx (hunk 5 above)
Comment (ccatherinetan) [reply src/app/(auth)/reset-password/page.tsx:58]: i'll leave it as is for now!

**Review by jinkang-0 — APPROVED** [review]
